iroh-resolver/src/unixfs_builder.rs

```diff
@@ -1,55 +1,172 @@
+        self.path = Some(path);
+        self
+    }
+
+    pub fn with_name(&mut self, name: impl Into<String>) -> &mut Self {
         self.name = Some(name.into());
         self
     }
 
+    pub fn add_dir(&mut self, dir: Directory) -> &mut Self {
+        let name = dir.name();
+        self.estimated_size += self.estimated_size + name.len() + CID_LEN;
+        if self.typ == DirectoryType::Basic
+            && self.estimated_size + BUFFER > self.chunker.chunk_size()
+        {
+            self.typ = DirectoryType::Hamt;
+        }
+        self.entries.push(Entry::Directory(dir));
+        self
+    }
+
     pub fn add_file(&mut self, file: File) -> &mut Self {
-        self.files.push(file);
+        let name = file.name();
+        self.estimated_size += self.estimated_size + name.len() + CID_LEN;
+        if self.typ == DirectoryType::Basic
+            && self.estimated_size + BUFFER > self.chunker.chunk_size()
+        {
+            self.typ = DirectoryType::Hamt;
+        }
+        self.entries.push(Entry::File(file));
         self
     }
 
-    pub async fn build(self) -> Result<Directory> {
-        let DirectoryBuilder { name, files } = self;
+    #[async_recursion(?Send)]
+    pub async fn build(mut self) -> Result<Directory> {
+        if let Some(path) = self.path {
+            let name = path
+                .file_name()
+                .and_then(|s| s.to_str())
+                .unwrap_or_default();
+
+            let mut builder = DirectoryBuilder::new();
+            builder.with_name(name);
+            let entries = std::fs::read_dir(path)?;
+            for entry in entries {
+                let entry = entry?;
+                let path = entry.path();
+                // TODO: this seems dumb, is this the only way to differentiate between
+                // files and dirs?
+                // TODO: what do we do with symlinks?
+                if path.is_file() {
+                    let mut file_builder = FileBuilder::new();
+                    file_builder.with_path(path);
+                    let f = file_builder.build().await?;
+                    builder.add_file(f);
+                } else if path.is_dir() {
+                    let mut dir_builder = DirectoryBuilder::new();
+                    dir_builder.with_path(path);
+                    let d = dir_builder.build().await?;
+                    builder.add_dir(d);
+                } else {
+                    anyhow::bail!("directory entry is neither file nor directory")
+                }
+            }
+            self = builder;
+        }
+
+        if self.typ == DirectoryType::Hamt {
+            // Question: is it too late to be erroring here? should we be
+            // erroring when we add the entry? to catch this as soon as
+            // possible?
+            anyhow::bail!("too many links to fit into one chunk, must be encoded as a HAMT. However, HAMT creation has not yet been implemented.");
+        }
```

Comment:
This is more of a question about changing the current API for DirectoryBuilder methods.

Should we be returning `Result` for `add_file` & `add_dir`, and erroring when we get to a hamt sized directory?